### ISSUE TITLE
New/different optimizations

### ISSUE THREAD

**Issue by ExtReMLapin** [JakuJ/BFC-10000#1]
Hello,
I wrote my own brainfuck transpiler with ti's own optimization passes and I ended up doing some of the similar optimizations you do

https://github.com/ExtReMLapin/fast_brainfuck.lua/blob/main/README.md

For the following optimization I use an Intermediate Representation system as it's much easier to work thru the multiple optimization passes.

# Assignation pass: 

You already do the clearing optimization.
If then look at the following brainfuck instruction it can become an assignation optimization as the next one could be for increasing the value of the current cell.

That's one step ahead of the clearing optimization pass. It doesn't give much performance boost but you'll need it for what comes next.

# Memset pass.

As you might end up with multiple assignations with the same integer (not increases but assignations , see above!) you can have one cell shift/tape position move between each assignation.

Which means you'll set the next N cells to the I value.
I turn this into a memset optimisation call : 

Example using lua pseudocode : 

![](https://i.imgur.com/tvvbs6j.png)

Also i'm not sure we're doing exactly the same thing on the loop multiplication unrolling, feel free to check my code on my repo

**Comment by JakuJ** [comment]
Hi @ExtReMLapin

I haven't done any work on repository for a long time, but I'll take a look at the optimisations you proposed.

As for the assignment pass, setting current cell's value to zero and then adding some value is probably optimised to a single assignment by LLVM at some point. I'm not sure though, I kinda mindlessly copied the LLVM optimisation passes from StackOverflow and I'm not quite sure what they do 😁 

Nevertheless you're right – having an assignment node in the AST would allow for memset-like optimisations in the future. I'll see what I can do.

As for the multiplication loops – I don't speak Lua, so I'll just try to explain what I do here. These are loops that begin with [- and then there's a series of alternating `move`s and `add`s, followed by moving back to the starting cell. It's a pattern used to multiply the cell's value by some number and store the result in another cell.

Consider this code: `+++[->>+++++>---<<<]`. We start on cell `i` with a value of 3 and then add multiples of it to other cells:  5 * 3 to cell `i + 2` and (-3) * 3 to cell `i + 3`.

After the optimisation, the AST for this code can thus be loosely translated to the following C pseudocode:

```c
tape[i] = 3;                           // +++
while (tape[i] != 0) {                 // [-
  tape[i + 2] += 5 * tape[i];          // >>+++++
  tape[i + 3] += (-3) * tape[i];       // >---
  tape[i] = 0;                         // <<<]
}
```

The implementation of this pass, given in the [MultiplicationLoopPass](https://github.com/JakuJ/BFC-10000/blob/master/src/MultiplicationLoopPass.cpp) file, is based on a Finite State Machine that accepts such loops. Attached is the diagram for this DFA. The '+' input symbol represents an `AddNode`, and the '>' symbol – a `MoveNode`. The F and H states are when the nodes corresponding to the `tape[i + offset] += mult * tape[i]` statements are added to the output code.

<img width="717" alt="Zrzut ekranu 2021-01-27 o 20 15 33" src="https://user-images.githubusercontent.com/23246747/106042782-cd19ef80-60dd-11eb-8c09-a4730192c792.png">

**Comment by JakuJ** [comment]
@ExtReMLapin I've added the optimisation passes you've mentioned. I've ran some basic benchmarks to see whether they actually brought noticeable perf boosts.

For reference, after compiling with `luajit <fast_brainfuck.lua> mandelbrot.bf out.lua` I ran it via `luajit <out.lua>` a few times and got ~3.1s. I used luajit v2.0.5.

As for the performance of the **compiled** Mandelbrot program:

Before:
* with LLVM optimization: ~1.15s
* without LLVM optimisation: ~3.33s

After:
* with LLVM optimisation: ~1.15s (unchanged)
* without LLVM optimisation: ~ 3.33s (unchanged)

Basically LLVM optimises these things regardless of the `-O` flag. 

As for the **interpreter**'s performance, it used to run the Mandelbrot in ~12s. Now it's ~9.2s, so I guess it's a success. 

Thanks for the ideas, I'm closing the issue now.
